stdlib/set.gr

```diff
@@ -0,0 +1,305 @@
+      }
+    }
+  }
+}
+
+export let remove = (key, set) => {
+  let buckets = set.buckets;
+  let idx = getBucketIndex(key, buckets);
+  let bucket = buckets[idx];
+  match (bucket) {
+    None => void,
+    Some(node) => {
+      // If it is a top-level node, just replace with next node
+      if (key == node.key) {
+        set.size = decr(set.size);
+        buckets[idx] := node.next;
+      } else {
+        if (removeInBucket(key, node)) {
+          set.size = decr(set.size);
+        }
+      }
+    }
+  }
+}
+
+export let size = (set) => {
+  set.size
+}
+
+export let isEmpty = (set) => {
+  size(set) == 0
+}
+
+export let clear = (set) => {
+  set.size = 0;
+  let buckets = set.buckets;
+  Array.forEachi((bucket, idx) => {
+    buckets[idx] := None;
+  }, buckets);
+}
+
+let rec forEachBucket = (fn, node) => {
+  match (node) {
+    None => void,
+    Some({ key, next }) => {
+      fn(key);
+      forEachBucket(fn, next);
+    }
+  }
+}
+
+export let forEach = (fn, set) => {
+  let buckets = set.buckets;
+  Array.forEach((bucket) => {
+    forEachBucket(fn, bucket)
+  }, buckets);
+}
+
+let rec reduceEachBucket = (fn, node, acc) => {
+  match (node) {
+    None => acc,
+    Some({ key, next }) =>
+      reduceEachBucket(fn, next, fn(acc, key))
+  }
+}
+
+export let reduce = (fn, init, set) => {
+  let buckets = set.buckets;
+  let mut acc = init;
+  Array.forEach((bucket) => {
+    acc = reduceEachBucket(fn, bucket, acc)
+  }, buckets);
+  acc
+}
+
+export let filter = (predicate, set) => {
+  let keysToRemove = reduce((list, key) =>
+    if (!predicate(key)) {
+      [key, ...list]
+    } else {
+      list
+    }, [], set);
+  List.forEach((key) => {
+    remove(key, set);
+  }, keysToRemove);
+}
+
+export let reject = (predicate, set) => {
+  filter((key) => !predicate(key), set)
+}
+
+export let toList = (set) => {
+  reduce((list, key) => [key, ...list], [], set)
+}
+
+export let fromList = (list) => {
+  let set = make();
+  List.forEach((key) => {
+    add(key, set);
+  }, List.unique(list));
+  set
+}
+
+export let toArray = (set) => {
+  Array.fromList(toList(set))
+}
+
+export let fromArray = (array) => {
+  let set = make();
+  Array.forEach((key) => {
+    add(key, set);
+  }, Array.unique(array));
+  set
+}
+
+export let union = (set1, set2) => {
+  let set = make();
+  forEach((key) => {
+    add(key, set);
+  }, set1)
+  forEach((key) => {
+    if (!contains(key, set1)) {
+      add(key, set);
+    }
+  }, set2);
+  set
+}
+
+export let diff = (set1, set2) => {
+  let set = make();
+  forEach((key) => {
+    if (!contains(key, set2)) {
+      add(key, set);
+    }
+  }, set1)
+  forEach((key) => {
+    if (!contains(key, set1)) {
+      add(key, set);
+    }
+  }, set2);
+  set
+}
+
+export let intersect = (set1, set2) => {
+  let set = make();
+  forEach((key) => {
+    if (contains(key, set2)) {
+      add(key, set);
+    }
+  }, set1)
+  forEach((key) => {
+    if (contains(key, set1) && !contains(key, set)) {
```

Comment:
Do we need the `!contains` or is that already handled by `add`? Which is faster?

Reply:
Thank you. Once again, I forgot that `add` should be sufficient.